{{cookiecutter.root_dir__update_if_you_intend_to_use_monorepo}}/{{cookiecutter.project_name}}/validation/model_validation_input.py

```diff
@@ -0,0 +1,108 @@
+##################################################################################
+# TODO : Please fill in parameters with proper values for your project
+##################################################################################
+import numpy as np
+from enum import Enum
+from mlflow.models import MetricThreshold, make_metric
+
+
+class RunMode(Enum):
+    ENABLED = 1
+    DRY_RUN = 2
+    DISABLED = 3
+
+
+# The `run_mode` defines whether model validation is enabled or not. It can be one of the three values:
+# - `DISABLED` : Do not run the model validation notebook.
+# - `DRY_RUN`  : Run the model validation notebook. Ignore failed model validation rules and proceed to move model to Production stage.
+# - `ENABLED`  : Run the model validation notebook. Move model to Production stage only if all model validation rules are passing.
+# TODO(required)
+def get_run_mode():
+    return RunMode.DISABLED
+
+
+# Whether to load the current registered "Production" stage model as baseline. A version with "Production" stage must
+# exist for the model.
+# Baseline model is a requirement for relative change and absolute change validation rules.
+# TODO(required)
+def enable_baseline_comparison():
+    return False
+
+
+# model validation data input for prod workspace. A Pandas DataFrame or Spark DataFrame, containing evaluation features and labels.
+# Please refer to data parameter in mlflow.evaluate documentation https://mlflow.org/docs/latest/python_api/mlflow.html#mlflow.evaluate
+# TODO(required)
+def get_prod_workspace_validation_input(spark):
```

Comment:
Feels weird to have separate prod and staging validation parameters in here.

Why isn't this a get_workspace_validation_input - this may become an issue for the other parameters, too.

I guess the question is how to expose these arguments cleanly (considering some of the arguments are not literalls, they can be python code).

Reply:
@vladimirk-db  Yeah I agree. 
There are several options for the API.
First question is which parameters will be shared by both staging and prod and which parameters are specific to one workspace. (If none of the parameters can be shared, we can just create two python input files)

My feeling is that the `model type`, `targets`, `custom metrics`, `thresholds` can be shared.
validation input `data` and `run mode` can not be shared.
`enable_baseline_comparison` maybe, maybe not.(always enable if there's a Production version?)

If only two parameters are different. we can either create separate methods(like I did).
Or pass Env as parameter, for example,
```
Class Env(Enum):
    DEV=1
    STAGING=2
    PROD=3

def get_workspace_validation_input(spark, env):
    if env == Env.PROD:
        return spark.sql(
            "SELECT * FROM delta.`dbfs:/databricks-datasets/nyctaxi-with-zipcodes/subsampled`"
        )
    elif env == Env.STAGING:
        return spark.sql(
            "SELECT * FROM delta.`dbfs:/databricks-datasets/nyctaxi-with-zipcodes/subsampled`"
        )
    return None

def get_workspace_run_mode(env)
    return RunMode.ENABLED if env == Env.PROD else RunMode.DISABLED
```

Or we can add the env parameter to all functions and let user decide. How you feel?

Reply:
Sorry for the delay. Thinking about this, we should be consistent in how we handle user-provided code + configs across all the components in Stacks (training, validation, deployment, feature engineering, monitoring).

I'd rather not introduce env-based input handling just for the validation step.

Maybe we can write a quick one decision doc about this, what do you think @mingyu89  (would be useful for  Stacks overall going forward IMHO).